plugins/inputs/influxdb_listener/influxdb_listener.go

```diff
@@ -271,45 +271,56 @@ func (h *InfluxDBListener) handleWrite() http.HandlerFunc {
 		parser := influx.NewStreamParser(body)
 		parser.SetTimeFunc(h.timeFunc)
 
-		var err error
 		precisionStr := req.URL.Query().Get("precision")
 		if precisionStr != "" {
-			var precision time.Duration
-			precision = getPrecisionMultiplier(precisionStr)
+			precision := getPrecisionMultiplier(precisionStr)
 			parser.SetTimePrecision(precision)
 		}
 
 		var m telegraf.Metric
+		var err error
 		var parseErrorCount int
-		var lastPos int
-		lastPos = 0
+		var lastPos int = 0
 		var firstParseErrorStr string
 		for {
+			select {
+			case <-req.Context().Done():
+				// Shutting down before parsing is finished.
+				res.WriteHeader(http.StatusServiceUnavailable)
+				return
+			default:
+			}
+
 			m, err = parser.Next()
 			pos := parser.Position()
 			h.bytesRecv.Incr(int64(pos - lastPos))
 			lastPos = pos
 
-			//continue parsing metrics even if some are malformed
+			// Continue parsing metrics even if some are malformed
 			if parseErr, ok := err.(*influx.ParseError); ok {
 				parseErrorCount += 1
 				errStr := parseErr.Error()
 				if firstParseErrorStr == "" {
 					firstParseErrorStr = errStr
-				} else {
 				}
 				continue
 			} else if err != nil {
+				// Either we're exiting cleanly (err ==
+				// influx.EOF) or there's an unexpected error
```

Comment:
do we want to log it if it's not influx.EOF?

Reply:
that's on line 323